File: cpp/src/oelMVCS/Framework.cpp
#include "oelMVCS/Core/Logger.h"
#include "oelMVCS/Core/Config.h"
#include "oelMVCS/Framework.h"

namespace XTC
{
namespace oelMVCS
{

StaticPipe* Framework::getStaticPipe()
{
    return staticPipe_;
}

DynamicPipe* Framework::getDynamicPipe()
{
    return dynamicPipe_;
}

void Framework::setConfig(const Config* _config)
{
    board_.config = _config;
}

void Framework::setLogger(const Logger* _logger)
{
    return logger_;
}

Framework::Framework()
{
    board_ = new Board();
    staticPipe_ = new StaticPipe(board_);
    dynamicPipe_ = new DynamicPipe(board_);
}

~Framework::Framework()
{
    if (0 != staticPipe_)
    {
        delete staticPipe_;
        staticPipe_ = 0;
    }
    if (0 != dynamicPipe_)
    {
        delete dynamicPipe_;
        dynamicPipe_ = 0;
    }
    if (0 != board_)
    {
        delete board_;
        board_ = 0;
    }
}

void Framework::Initialize()
{
    board_->getLogger()->Info("initialize framework");
}

void Framework::Setup()
{
    board_->getLogger()->Info("setup framework");
    board_->getViewCenter()->Setup();
    board_->getServiceCenter()->Setup();
    board_->getModelCenter()->Setup();
    board_->getControllerCenter()->Setup();
}

void Framework::Dismantle()
{
    board_->getLogger()->Info("dismantle framework");
    board_->getViewCenter()->Dismantle();
    board_->getServiceCenter()->Dismantle();
    board_->getModelCenter()->Dismantle();
    board_->getControllerCenter()->Dismantle();
}


void Framework::Release()
{
    board_->getLogger()->Info("release framework");
}

}//namespace
}//namespace

File: csharp/oelMVCS/Framework.cs
/********************************************************************
     Copyright (c) XTechCloud
     All rights reserved.
*********************************************************************/

using System.Collections.Generic;

namespace XTC.oelMVCS
{
    /// <summary>
    /// 框架类
    /// </summary>
    public class Framework
    {
        private StaticPipe staticPipe_;
        private DynamicPipe dynamicPipe_;
        private Board board_;

        /// <summary>
        /// 用户数据映射表
        /// </summary>
        private Dictionary<string, UserData> userDatas_ = new Dictionary<string, UserData>();

        #region Get Set
        public StaticPipe getStaticPipe()
        {
            return staticPipe_;
        }

        public DynamicPipe getDynamicPipe()
        {
            return dynamicPipe_;
        }

        public void setConfig(Config _value)
        {
            board_.setConfig(_value);
        }

        public void setLogger(Logger _value)
        {
            board_.setLogger(_value);
        }
        #endregion

        public Framework()
        {
            board_ = new Board();
            staticPipe_ = new StaticPipe(board_);
            dynamicPipe_ = new DynamicPipe(board_);
        }

        ~Framework()
        {
        }

        /// <summary>
        /// 设置用户数据
        /// </summary>
        /// <param name="_key">键</param>
        /// <param name="_value">值，为空时删除对应的键</param>
        public void setUserData(string _key, UserData? _value)
        {
            if(null == _value)
            {
                if (userDatas_.ContainsKey(_key))
                    userDatas_.Remove(_key);
                return;
            }
            userDatas_[_key] = _value;
        }

        /// <summary>
        /// 获取用户数据
        /// </summary>
        /// <param name="_key">键</param>
        /// <returns>不存在时返回空</returns>
        public UserData? getUserData(string _key)
        {
            if (!userDatas_.ContainsKey(_key))
                return null;
            return userDatas_[_key];
        }


        /// <summary>
        /// 框架初始化，完成各层中心的实例化
        /// </summary>
        public void Initialize()
        {
            board_.getLogger()?.Info("initialize framework");
            board_.setModelCenter(new ModelCenter(board_));
            board_.setViewCenter(new ViewCenter(board_));
            board_.setControllerCenter(new ControllerCenter(board_));
            board_.setServiceCenter(new ServiceCenter(board_));
        }

        /// <summary>
        /// 框架安装，完成各层中心已注册组件的安装
        /// 此过程将调用各派生组件的setup方法
        /// </summary>
        public void Setup()
        {
            board_.getLogger()?.Info("setup framework");
            board_.getModelCenter().PreSetup();
            board_.getViewCenter().PreSetup();
            board_.getControllerCenter().PreSetup();
            board_.getServiceCenter().PreSetup();
            board_.getModelCenter().Setup();
            board_.getViewCenter().Setup();
            board_.getControllerCenter().Setup();
            board_.getServiceCenter().Setup();
            board_.getModelCenter().PostSetup();
            board_.getViewCenter().PostSetup();
            board_.getControllerCenter().PostSetup();
            board_.getServiceCenter().PostSetup();
        }

        /// <summary>
        /// 框架拆卸，完成各层中心已注册组件的拆卸
        /// 此过程将调用各派生组件的dismantle方法
        /// </summary>
        public void Dismantle()
        {
            board_.getLogger()?.Info("dismantle framework");
            board_.getServiceCenter().PreDismantle();
            board_.getControllerCenter().PreDismantle();
            board_.getViewCenter().PreDismantle();
            board_.getModelCenter().PreDismantle();
            board_.getServiceCenter().Dismantle();
            board_.getControllerCenter().Dismantle();
            board_.getViewCenter().Dismantle();
            board_.getModelCenter().Dismantle();
            board_.getServiceCenter().PostDismantle();
            board_.getControllerCenter().PostDismantle();
            board_.getViewCenter().PostDismantle();
            board_.getModelCenter().PostDismantle();
        }


        /// <summary>
        /// 框架销毁，完成各层中心的释放
        /// </summary>
        public void Release()
        {
            board_.getLogger()?.Info("release framework");
            board_.setViewCenter(null);
            board_.setServiceCenter(null);
            board_.setModelCenter(null);
            board_.setControllerCenter(null);
        }

    }

}//namespace
